htlcinterceptor/middleware.go

```diff
@@ -0,0 +1,146 @@
+		holdForwards: make(map[channeldb.CircuitKey]ForwardResolver),
+	}
+}
+
+// AddForwardHtlcHandler adds a ForwardHtlcHandler.
+func (c *Middleware) AddForwardHtlcHandler(
+	handler ForwardHtlcHandler) uint64 {
+
+	id := atomic.AddUint64(&c.fwdHandlerID, 1)
+
+	c.Lock()
+	c.fwdHandlers[id] = handler
+	c.Unlock()
+
+	// Return the id so that a caller can call RemoveForwardHtlcHandler.
+	return id
+}
+
+// RemoveForwardHtlcHandler removes a ForwardHtlcHandler.
+func (c *Middleware) RemoveForwardHtlcHandler(id uint64) {
+	c.Lock()
+	delete(c.fwdHandlers, id)
+	c.Unlock()
+}
+
+// InterceptForwardHtlc is called before each forward and query its registered
+// handlers to see if one of them is interested in handling (hold) this
+// forward. In case one of them returns true the forward will be marked as hold
+// and signals the switch to wait for a later resolve action which should be
+// done later by calling one of:
+// * ResumeHoldForward
+// * SettleHoldForward
+// * FailHoldForward.
+func (c *Middleware) InterceptForwardHtlc(
+	circuitKey channeldb.CircuitKey,
+	htlc lnwire.UpdateAddHTLC, resolver ForwardResolver) bool {
+
+	c.Lock()
+	defer c.Unlock()
+	for _, actionResolver := range c.fwdHandlers {
+		if actionResolver(circuitKey, htlc) {
+			c.holdForwards[circuitKey] = resolver
+			return true
+		}
+	}
+
+	// None of the resolvers handled the forward.
+	return false
+}
+
+// ResumeHoldForward notifies the intention to resume an existing hold forward.
+// This basically means the caller wants to resume with the default behavior
+// for this htlc which usually means forward it. If an existing hold forward
+// doesn't exist it returns an error.
+func (c *Middleware) ResumeHoldForward(
+	circuitKey channeldb.CircuitKey) error {
+
+	return c.resolveHoldForward(circuitKey,
+		func(resolver ForwardResolver) error {
+			return resolver.Resume()
+		})
+}
+
+// SettleHoldForward notifies the intention to settle an existing hold forward
+// with a given preimage. If an existing hold forward doesn't exist it returns
+// an error.
+func (c *Middleware) SettleHoldForward(
+	circuitKey channeldb.CircuitKey, preimage lntypes.Preimage) error {
+
+	return c.resolveHoldForward(circuitKey,
+		func(resolver ForwardResolver) error {
+			return resolver.Settle(preimage)
+		})
+}
+
+// FailHoldForward notifies the intention to fail an existing hold forward. If
+// an existing hold forward doesn't exist it returns an error.
+func (c *Middleware) FailHoldForward(
+	circuitKey channeldb.CircuitKey) error {
+
+	return c.resolveHoldForward(circuitKey,
+		func(resolver ForwardResolver) error {
+			return resolver.Fail()
+		})
+}
+
+func (c *Middleware) resolveHoldForward(
+	circuitKey channeldb.CircuitKey, resolveFunc func(ForwardResolver) error) error {
+
+	c.Lock()
+	resolver, ok := c.holdForwards[circuitKey]
+	c.Unlock()
+	if !ok {
+		return ErrFwdNotExists
+	}
+	if err := resolveFunc(resolver); err != nil {
+		return err
+	}
+
+	delete(c.holdForwards, circuitKey)
```

Comment:
shoudn't this be protected by the lock too? concurrent writes to maps are not supported

Reply:
We should! Done.